### PR TITLE
[decomp] jak2: sprite, sprite-distort, some of sprite-glow

### PR DESCRIPTION
**Pull request by Francessco121**
- `sprite` - Basically the same as Jak 1
  - `sprite-frame-data` changed a little
  - Shadow sprite stuff is gone
  - New stuff for glow sprites
  - VU program is a little different but nothing major
- `sprite-distort` - Identical except...
  - Doesn't take a ptr to the current framebuffer anymore and instead hardcodes the texture ptr that is read from
  - No longer takes into account the screen height
  - VU program is identical
- `sprite-glow` - Couldn't finish because...
  - It depends on `adgif-shader<-texture-simple!` from `texture`
  - Couldn't figure out the parameters for `sprite-glow-add-simple-sprite` (need to find something that calls it)

### CODE REVIEW

**Comment by ManDude** [comment]
RIP PS1 shadow

File: goal_src/jak2/engine/gfx/sprite/sprite.gc [hunk 1]
```diff
@@ -5,5 +5,1073 @@
...
+    (set! (-> out w) (sqrtf (- (- (- 1.0 (* f2-0 f2-0)) (* f1-0 f1-0)) (* f0-0 f0-0))))
+    )
+  out
+  )
+
+(defun sprite-add-matrix-data ((dma-buff dma-buffer) (matrix-mode uint))
+  "Add matrix data to a dma-buffer.
+   Mode 0 = send (-> *math-camera* camera-temp)
+   Mode 1 = screen sprites, like the HUD"
+  (let ((count 900)) ;; maybe address??
```
Comment (water111) [inline goal_src/jak2/engine/gfx/sprite/sprite.gc:588]: yeah, I think this is the address for where the matrix goes. See `SpriteDataMem::Matrix = 900` in `sprite_common.h`.

File: goal_src/jak2/engine/gfx/sprite/sprite.gc [hunk 2]
```diff
@@ -5,5 +5,1073 @@
...
+              (t0-4 (-> t1-0 vector 2 quad))
+              (t1-1 (-> t1-0 trans quad))
+              )
+         (set! (-> mtx vector 0 quad) a2-4)
+         (set! (-> mtx vector 1 quad) a3-4)
+         (set! (-> mtx vector 2 quad) t0-4)
+         (set! (-> mtx trans quad) t1-1)
+         )
+       (&+! (-> dma-buff base) 64)
+       ;; also send this thing.
```
Comment (water111) [inline goal_src/jak2/engine/gfx/sprite/sprite.gc:616]: `hvdf-off` = Horizontal, Vertical, Depth, Fog Offsets.
After the perspective transformation, this value is added the vector, giving the final horizontal, vertical, depth, and fog values.

Comment (water111) [reply goal_src/jak2/engine/gfx/sprite/sprite.gc:616]: As a side note, they actually do fogging correctly, and don't fog based on depth, but instead actual distance from the camera.

File: test/decompiler/vu_reference/jak2/sprite-glow-result.txt [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  sq.xy vf12, 79(vi05)       |  nop                      
+  sq.xy vf13, 81(vi05)       |  nop                      
+  sq.xy vf14, 83(vi05)       |  nop                      
+  xgkick vi05                |  nop                      
+L2:
+  iaddiu vi01, vi00, 0x694   |  nop                      
+  ibne vi00, vi03, L1        |  nop                      
+  isub vi05, vi01, vi05      |  nop                      
+  nop                        |  nop :e                   
+  nop                        |  nop
```
Comment (water111) [inline test/decompiler/vu_reference/jak2/sprite-glow-result.txt:136]: It's nice to see that this isn't too long.

**Review by water111 — APPROVED** [review]
:+1: 
Awesome documentation!

I'm glad to see that there's not too much new from Jak 1. I suspect that we can reuse a lot of code on the C++ side.

File: goal_src/jak2/engine/gfx/sprite/sprite.gc (hunk 1 above)
Comment (Francessco121) [reply goal_src/jak2/engine/gfx/sprite/sprite.gc:588]: 🙂 for what it's worth, i think these are yours or someone else's comments. i copied everything from the jak 1 code that still made sense and tweaked a little

**Pull request merged**